File: Problems/141. Linked List Cycle.java
import java.util.Set;
import java.util.HashSet;

 class ListNode {
      int val;
      ListNode next;
      ListNode(int x) {
          val = x;
          next = null;
      }
  }

  class Solution {

      private Set<ListNode> set = new HashSet<>();
      public boolean hasCycle(ListNode head) {

          if(head == null){
              return false;
          }

          ListNode cur = head;

          while(true){

              if(set.contains(cur)){
                  return true;
              }

              set.add(cur);
              if(cur.next == null){
                  break;
              }
              cur = cur.next;
          }

        return false;
    }
}

File: 141. Linked List Cycle.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {

	if head == nil || head.Next == nil {
		return false
	}

	slow := head
	fast := head.Next

	for fast != nil && fast.Next != nil {
		if fast == slow {
			return true
		}
		slow = slow.Next
		fast = fast.Next.Next
	}

	return false
}

func main() {

	l1 := &ListNode{3, nil}
	l2 := &ListNode{2, nil}
	l3 := &ListNode{0, nil}
	l4 := &ListNode{-4, nil}

	l1.Next = l2
	l2.Next = l3
	l3.Next = l4
	l4.Next = l2

	answer := hasCycle(l1)

	fmt.Println("ANswer: ", answer)
}
